### PR TITLE
Tentative solution for improving our current network and SSL setup

### PR DESCRIPTION
**Pull request by BjoernKW**
https://github.com/stratospheric-dev/stratospheric/issues/59
https://github.com/stratospheric-dev/stratospheric/issues/61

### CODE REVIEW

**Comment by rieckpil** [comment]
Thanks for taking the time to come up with a possible solution 👍

I'd personally prefer to solve this issue with a lasting solution over a tentative solution. Many readers already struggle here and might lose the motivation to keep going.

I guess the main decision we have to make is on what to focus on with our CDK stacks: supporting incremental iterations that might include various `if-else` constructs in our stacks vs. strict stack setups that support only a specific use case.

I can see that the SSL setup is maybe too much overhead at the beginning of the book as we want readers to have a successful `ServiceStack` running early on. We have to pass a custom certificate whenever we want to create an HTTPS listener for our ELB. AWS [doesn't create one](https://stackoverflow.com/questions/68567684/why-does-awss-application-load-balancer-require-a-custom-domain-when-using-http) for us for the ELB domain. 

However, we currently have also a small inconsistency in the flow of the book as e.g. we already require the `applicationUrl` for the Cognito stack and the Canary stack before we talk about setting up the custom domain. The more I think about this the more I assume we have to restructure the book and explain the SSL + domain step at the beginning. If a reader deploys the Network stack without a `sslCertificateArn` and uses a custom domain, there's no way to get past the Cognito chapter (if they plan to build the infrastructure in parallel to reading the book).

I can see two scenarios (or personas) for our `NetworkStack` & `ServiceStack`:

1. I'm reading the book and want to try out things on my own. I'd like to see the Hello World Spring Boot example running on AWS using the core CDK infrastructure (network + service stack, maybe ECR). I don't care about security and want a quick win early on. Hence I'm fine with only an HTTP Listener to see the app running in the cloud. I don't care about SSL and a custom domain as I'm just playing around and learning.
2. I want to deploy the entire CDK infrastructure for a production use case. Hence, security and a custom domain are important to me. I'm willing to take the extra effort and create an SSL cert and a custom domain. 

So my proposed solution would be:

1. Within our `NetworkStack` we decide which ELB Listeners to create and only support two scenarios:
   1. No `sslCertificateArn` passed -> only create HTTP listener to support the `SpringBootApplicationStack` use case
   2. `sslCertificateArn` passed -> create HTTP and HTTPS listener and let the HTTP listener redirect to the HTTPS listener
3. Everything past chapter 5. requires an SSL cert and a custom domain. We somehow have to move and integrate chapter 19 `Configuring HTTPS and a Custom Domain with Route 53 and ELB` when demonstrating the `NetworkApp` in chapter 6.

Otherwise, it's hard to follow the book step-by-step without too many `if-else` conditions inside our stacks.

WDYT @thombergs, @BjoernKW?

**Comment by BjoernKW** [comment]
@rieckpil I like the idea. If I understand this correctly, this just means moving the HTTP->HTTPS redirect setup from the `DomainStack` in the application's CDK project to the `Network` stack in https://github.com/stratospheric-dev/cdk-constructs .

I've just changed the branches for the two pull requests accordingly. Please have a look at the changes and see if that's what you had in mind.

If so, this would require (minimal) changes to this chapters from the book as well (mainly updating the file references and a short explanation as to why we're already adding this behaviour in the `Network` stack, which should amount to moving that section from the "HTTPS and Custom Domain" chapter to "Designing a Deployment Project with CDK").

* Designing a Deployment Project with CDK
* Configuring HTTPS and a Custom Domain with Route 53 and ELB

**Comment by lealceldeiro** [comment]
@rieckpil would you consider a third scenario, for the CDK setup?:

3. I want to deploy the complete infrastructure in "iterations" (or incrementally), as my very small devs group moves on with our app implementation, hence I may have a (sub) domain and want to start trying things out, but without the hassle (yet) of creating a SSL Certificate (so: domain ready, but not ssl certificate -- we're intending to move to production at some point, but we're fine testing without ssl certificate/security in staging or dev environment in AWS).

To me, this is a scenario previous to your scenario number 2:

> I want to deploy the entire CDK infrastructure for a production use case. Hence, security and a custom domain are important to me. I'm willing to take the extra effort and create an SSL cert and a custom domain.

**Comment by rieckpil** [comment]
Thanks for the third suggestion @lealceldeiro 👍 

From a maintenance perspective for this eBook, I'd avoid supporting this third scenario officially. The main reason is the additional (almost business) logic that we would have to embed into our stacks on what to create.

If you have your Domain ready, creating the SSL certificate is a matter of seconds if you use Route53. If you don't use Route53, the only manual step you'd have to do is to add the verification lines manually to your Host file. 

Nevertheless, you can still build such a CDK construct using our eBook as a source of inspiration.

**Review by rieckpil — APPROVED** [review]
@BjoernKW that's exactly what I meant 👍
